src/garage/torch/algos/dqn.py

```diff
@@ -0,0 +1,268 @@
+            clip_rewards]. If None, rewards are not clipped.
+        clip_gradient (float): Clip gradient norm to `clip_gradient`. If None,
+            gradient are not clipped. Defaults to 10.
+        reward_scale (float): Reward scale.
+    """
+
+    def __init__(
+            self,
+            env_spec,
+            policy,
+            qf,
+            replay_buffer,
+            exploration_policy=None,
+            eval_env=None,
+            qf_optimizer=torch.optim.Adam,
+            *,  # Everything after this is numbers.
+            steps_per_epoch=20,
+            n_train_steps=50,
+            max_episode_length_eval=None,
+            buffer_batch_size=64,
+            min_buffer_size=int(1e4),
+            discount=0.99,
+            qf_lr=_Default(1e-3),
+            clip_rewards=None,
+            clip_gradient=10,
+            target_update_freq=5,
+            reward_scale=1.):
+        self._clip_reward = clip_rewards
+        self._clip_grad = clip_gradient
+
+        self._steps_per_epoch = steps_per_epoch
+        self._target_update_freq = target_update_freq
+        self._episode_qf_losses = []
+        self._epoch_ys = []
+        self._epoch_qs = []
+
+        self._policy = policy
+        self._qf = qf
+        self._n_train_steps = n_train_steps
+
+        self._min_buffer_size = min_buffer_size
+        self._qf = qf
+        self._steps_per_epoch = steps_per_epoch
+        self._n_train_steps = n_train_steps
+        self._buffer_batch_size = buffer_batch_size
+        self._discount = discount
+        self._reward_scale = reward_scale
+        self.max_episode_length = env_spec.max_episode_length
+        self._max_episode_length_eval = (max_episode_length_eval
+                                         or self.max_episode_length)
+        self._episode_reward_mean = collections.deque(maxlen=100)
+
+        self.env_spec = env_spec
+        self.replay_buffer = replay_buffer
+        self.policy = policy
+        self.exploration_policy = exploration_policy
+
+        self._target_qf = copy.deepcopy(self._qf)
+        self._qf_optimizer = make_optimizer(qf_optimizer,
+                                            module=self._qf,
+                                            lr=qf_lr)
+        self._eval_env = eval_env
+        self.worker_cls = FragmentWorker
+
+    def train(self, trainer):
+        """Obtain samplers and start actual training for each epoch.
+
+        Args:
+            trainer (Trainer): Experiment trainer.
+
+        Returns:
+            float: The average return in last epoch cycle.
+
+        """
+        if not self._eval_env:
+            self._eval_env = trainer.get_env_copy()
+        last_returns = [float('nan')]
+        trainer.enable_logging = False
+
+        for _ in trainer.step_epochs():
+            for cycle in range(self._steps_per_epoch):
+                trainer.step_path = trainer.obtain_episodes(trainer.step_itr)
```

Comment:
Thanks for this implementation! I tried it out and noticed the exploration policy wasn't logging since the update method isn't called. In the TF implementation this line is followed by:

```python              
if hasattr(self.exploration_policy, 'update'):
   self.exploration_policy.update(trainer.step_path)
```

Reply:
Ah nice catch thank you!